### PR TITLE
STYLE: Avoid old-style cast

### PR DESCRIPTION
**Pull request by Leengit**
Replace old-style cast `(int)a` or `int(a)` with new-style `static_cast<int>(a)`, etc.

Avoiding old-style casts on advice of:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es48-avoid-casts
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es49-if-you-must-use-a-cast-use-a-named-cast
https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines-pro-type-cstyle-cast.html

Thank you @N-Dekker for the references.

## PR Checklist
- [X] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [X] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [X] Added test (or behavior not changed)
- [X] Updated API documentation (or API not changed)
- [X] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [X] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [X] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

### CODE REVIEW

File: Modules/Core/Common/include/itkThreadSupport.h
Comment (dzenanz) [inline Modules/Core/Common/include/itkThreadSupport.h:65]: This change might be a problem.

**Review by dzenanz — COMMENTED** [review]
Mostly looks good.

File: Modules/Core/Common/include/itkThreadSupport.h
Comment (dzenanz) [reply Modules/Core/Common/include/itkThreadSupport.h:65]: Yes, it is a problem: `D:\a\1\s\Modules\Core\Common\include\itkThreadSupport.h(65): error C2059: syntax error: 'static_cast'`.

File: Modules/Core/Common/include/itkFloodFilledSpatialFunctionConditionalConstIterator.hxx
Comment (N-Dekker) [inline Modules/Core/Common/include/itkFloodFilledSpatialFunctionConditionalConstIterator.hxx:115]: Thanks  @Leengit  I'm glad to see this `static_cast<double>(2.0)` being simplified to just `2.0` 😄   Note that with C++11/C++14, the other argument `static_cast<double>(dim)` should not need to be casted either, anymore, as the general form `pow(Arithmetic1 base, Arithmetic2 exp)` should be supported now with C++: https://en.cppreference.com/w/cpp/numeric/math/pow  However, we might as well defer further simplification of `std::pow` calls to another PR.

Comment (N-Dekker) [reply Modules/Core/Common/include/itkFloodFilledSpatialFunctionConditionalConstIterator.hxx:115]: Honestly I think it would be clearer to keep manual removal of redundant/unnecessary existing static_cast's separate from all those code changes that add static_cast's (in a separate commit, or even a separate PR). To just have a commit that only deals with old-style C-casts.

File: Modules/Core/Common/include/itkThreadSupport.h
Comment (Leengit) [reply Modules/Core/Common/include/itkThreadSupport.h:65]: Thank you.  I will fix that.

File: Modules/Core/Common/include/itkFloodFilledSpatialFunctionConditionalConstIterator.hxx
Comment (Leengit) [reply Modules/Core/Common/include/itkFloodFilledSpatialFunctionConditionalConstIterator.hxx:115]: @N-Dekker: Yes, I agree with both comments.  As far as changing this pull request to be consistent with them ... getting there from here would be a decent amount of work.  Is it good enough as is?

Comment (N-Dekker) [reply Modules/Core/Common/include/itkFloodFilledSpatialFunctionConditionalConstIterator.hxx:115]: @Leengit Didn't you run a script to do most of the job?

Comment (Leengit) [reply Modules/Core/Common/include/itkFloodFilledSpatialFunctionConditionalConstIterator.hxx:115]: Yes, a few `sed` commands.  Then fixed a few things by hand that my `sed` work got wrong.  Then more `sed` commands when I figured out whole categories that could be better.  Then more manual work to correct some more cases.

Comment (Leengit) [reply Modules/Core/Common/include/itkFloodFilledSpatialFunctionConditionalConstIterator.hxx:115]: The changes in this pull request from `T(x)` and `(T)x` are only for instances of `T` that are basic C++ types.  If there are cases that would benefit from similar changes, but where `T` is a class or template, those could be addressed in another pull request.

**Comment by Leengit** [comment]
I switched this to a draft pull request pending further investigation.

File: Modules/Core/Common/include/itkFloodFilledSpatialFunctionConditionalConstIterator.hxx
Comment (N-Dekker) [reply Modules/Core/Common/include/itkFloodFilledSpatialFunctionConditionalConstIterator.hxx:115]: I'm very interested to see those `sed` commands. Maybe I can use them myself in another project  😸

Comment (Leengit) [reply Modules/Core/Common/include/itkFloodFilledSpatialFunctionConditionalConstIterator.hxx:115]: Here they are in their ugliness as I used them.  Huge disclaimer: several of these did too little or too much and had to be corrected with subsequent commands.  And, in between many of them, there were edits by hand.  In short, these require careful handling and much work, and will still need fixes by hand.  You have been warned.  :smiley:

If you turn this into something less one-off, I'd be very interested to see it.

```sh
# It is important to use the top-level directory for the `git diff --name-only` commands
cd ~/git/ITK
# I should have used this alias more, from the start
alias findITKCXXFiles='(cd ~/git/ITK ; find * -type f | egrep -i -e '"'"'\.[it]?(cc|hh|[ch](\+\+|pp?|xx)?)$'"'"' | fgrep -v ThirdParty)'

egrep -lR '\((float|double|bool|char|short|int|long|unsigned| )+\)\(' | xargs sed -i -r -e 's/\((float|double|bool|char|short|int|long|unsigned| )+\)\(/static_cast<\1>(/g'
egrep -lR '\((float|double|bool|char|short|int|long|unsigned| )+\)\(' | xargs sed -i -r -e 's/\(((float|double|bool|char|short|int|long|unsigned| )+)\)\(/static_cast<\1>(/g'
findITKCXXFiles | xargs sed -i -r -e 's/double\(/static_cast<double>(/g'
git diff --name-only | xargs sed -i -r -e 's/rstatic_cast<double>/rdouble/g'
findITKCXXFiles | xargs sed -i -r -e 's/([^A-Za-z0-9_])float\(/\1static_cast<float>(/g'
findITKCXXFiles | xargs sed -i -r -e 's/([^A-Za-z0-9_])((float|double|bool|char|short|int|long|unsigned| )+)\(/\1static_cast<\2>(/g'
findITKCXXFiles | xargs sed -i -r -e 's/([^A-Za-z0-9_])((float|double|bool|char|short|int|long|unsigned)( (float|double|bool|char|short|int|long|unsigned))*)\(/\1static_cast<\2>(/g'
git diff --name-only | xargs sed -i -r -e 's/operator static_cast<bool>/operator bool/g'
git diff --name-only | xargs sed -i -r -e 's/return static_cast<bool>/return bool/g'
findITKCXXFiles | xargs sed -i -r -e 's/static_cast<float>\(([0-9\.]+)\)/\1f/g'
findITKCXXFiles | xargs sed -i -r -e 's/static_cast<float>\(([0-9]+)(\.([0-9]*))?\)/\1.\3f/g'
git diff --name-only | xargs sed -i -r -e 's/([0-9]+)\.f/\1.0f/g'
git diff master --name-only | xargs sed -i -r -e 's/std::function<static_cast<float>(float)>/std::function<float(float)>/g'
git diff master --name-only | xargs sed -i -r -e 's/std::function<static_cast<float>\(float\)>/std::function<float(float)>/g'
git diff master --name-only | xargs sed -i -r -e 's/std::function<static_cast<float>/std::function<float/g'
git diff master --name-only | xargs sed -i -r -e 's/std::function<static_cast<double>/std::function<double/g'
git diff master --name-only | xargs sed -i -r -e 's/std::function<static_cast<double>/std::function<double/g'
findITKCXXFiles | xargs sed -i -r -e  's/\(((float|double|bool|char|short|int|long|unsigned|const)( (float|double|bool|char|short|int|long|unsigned|const))*)\)((\[|\(|[->A-Za-z0-9_:])(\[|\]|\(|\)|[->\.A-Za-z0-9_:])*)/static_cast<\1>(\5)/g'
sed -i -r -e 's/static_cast < int\(*\) static_cast<int>\(> \(/static_cast<int (*)(int)>(/g'
sed -i -r -e 's/static_cast < int\(*\) static_cast<int>\(> \(/static_cast<int (*)(int)>(/g' Modules/IO/GDCM/src/itkGDCMImageIO.cxx
findITKCXXFiles | xargs sed -i -r -e 's/static_cast<double>\((-?[0-9]+\.[0-9]+([eE][0-9]+)?)\)/\1/g'
```

File: Modules/Core/Common/test/itkMetaDataDictionaryGTest.cxx
Comment (N-Dekker) [inline Modules/Core/Common/test/itkMetaDataDictionaryGTest.cxx:37]: Sorry, this seems beyond the scope of this commit.

File: Examples/Statistics/NormalVariateGenerator.cxx
Comment (Leengit) [reply Examples/Statistics/NormalVariateGenerator.cxx]: Done, via a distinct commit.

File: Modules/Core/Common/test/itkMetaDataDictionaryGTest.cxx
Comment (Leengit) [reply Modules/Core/Common/test/itkMetaDataDictionaryGTest.cxx:37]: But you just  requested this sort of translation for `int`, so why not `float`?  Is there a case where this textual substitution could fail?

Comment (N-Dekker) [reply Modules/Core/Common/test/itkMetaDataDictionaryGTest.cxx:37]: This is somewhere in the middle of what looks like an automatic "bulk" of ~1000 changes that mostly add ` static_cast`'s. In order to approve the PR, I would have to check all those lines.

File: Examples/Segmentation/HoughTransform2DLinesImageFilter.cxx
Comment (Leengit) [reply Examples/Segmentation/HoughTransform2DLinesImageFilter.cxx:286]: I did it in only `static_cast` situations precisely to avoid having this pull request wander into other territory.  It could get complicated in a wider scope and I didn't want to attempt that here.  However, I can undo any of the commits; please let me know.

Comment (N-Dekker) [reply Examples/Segmentation/HoughTransform2DLinesImageFilter.cxx:286]: My preference would be to have a separate PR dealing with all those "optional ints", not just in the context of casting. Which could be either before or after this PR. It might trigger some discussion about style. But it would be clearly a separate discussion from the casting issue.

Comment (Leengit) [reply Examples/Segmentation/HoughTransform2DLinesImageFilter.cxx]: That works for me.  I have removed the commit with this change.

**Comment by N-Dekker** [comment]
@Leengit Could you possibly extend the commit message text a little bit? It makes so many changes to a lot of previous work, that it might need a little explanation, or even justification.

You might refer to https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es48-avoid-casts or https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es49-if-you-must-use-a-cast-use-a-named-cast which warn about casts, and especially about C-style casts. Or maybe mention that such casts would trigger Clang-Tidy warnings, when enabled: https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines-pro-type-cstyle-cast.html

File: Modules/Core/Common/test/itkMetaDataDictionaryGTest.cxx
Comment (Leengit) [reply Modules/Core/Common/test/itkMetaDataDictionaryGTest.cxx:37]: I will revert these with the next commit.

**Comment by Leengit** [comment]
> Could you possibly extend the commit message text a little bit?

Done.

**Comment by Leengit** [comment]
Would someone with sufficient permission re-launch the failing check please

**Comment by N-Dekker** [comment]
It would be interesting to see if this pull request does indeed fix all possible Clang-Tidy C-style cast warnings (https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines-pro-type-cstyle-cast.html). Not a request, just in case you have some more time  😸 BTW, MSVC does have a similar Code Analysis check, saying "warning C26493, Don't use C-style casts"

**Comment by Leengit** [comment]
> Clang-Tidy C-style cast warnings

If `clang-tidy` can find them, it is a shame that `clang-format` won't make these edits for us.

**Comment by N-Dekker** [comment]
> If `clang-tidy` can find them, it is a shame that `clang-format` won't make these edits for us.

 `clang-tidy` is actually able to find C-style casts _and_ fix them, replacing them by static_cast, just like you did! I did not know that before, but I just tried it out, in a minimal example.

https://clang.llvm.org/extra/clang-tidy/checks/google-readability-casting.html says:

> This check is similar to -Wold-style-cast, but it suggests automated fixes in some cases.

**Comment by Leengit** [comment]
> `clang-tidy` is actually able to find C-style casts _and_ fix them

@N-Dekker I am quite tired of this pull request.  If you can make use of `clang-tidy` in another pull request, I'd gladly defer to that.

**Comment by N-Dekker** [comment]
> @N-Dekker I am quite tired of this pull request. 

 @Leengit Sorry if I made you tired! I did not mean to! While I did criticize your pull request, overall, I think it is very much an improvement of the code base. Wouldn't it be useful to have it merged as it is right now, and then possibly do some clang-tidy checks afterwards?

> If you can make use of clang-tidy in another pull request, I'd gladly defer to that.

Honestly my clang-tidy skills are just so-so. Running inside Visual Studio, using Clang Power Tools, it isn't very stable on my machine. So I'm not sure if I can. Ideally I guess it should run on the command-line instead. But I don't know how exactly. Do you?

**Comment by Leengit** [comment]
@N-Dekker Your feedback is always welcome; please keep up the good work!  I haven't tried `clang-tidy` in any context.  If you won't be pursuing that either then, yes, lets merge this pull request once it passes its checks.

**Comment by Leengit** [comment]
I just fixed a bug that only MacOS caught.  I am marking this pull request as ready for review; and it can be merged if it now passes its checks.

**Review by N-Dekker — APPROVED** [review]

**Comment by Leengit** [comment]
Please relaunch the `ITK.macOS.Python` check.  The error there is `##[error]We stopped hearing from agent Azure Pipelines 2. Verify the agent machine is running and has a healthy network connection. Anything that terminates an agent process, starves it for CPU, or blocks its network access can cause this error. For more information, see: https://go.microsoft.com/fwlink/?linkid=846610`

I hesitate to ignore this check because this check is the one that found the last real error, and I'd like to see it go all green.

**Comment by Leengit** [comment]
I just did a force-push to rebase to the current master.  That has the side effect of relaunching `ITK.macOS.Python`, so let's see. how that goes.

**Comment by N-Dekker** [comment]
Thanks @Leengit. In the meantime, I still haven't figured out how such a style improvement could be done automatically. I wish it could be something like:
    
    clang-tidy.exe -checks=google-readability-casting -fix -solution=ITK.sln -configuration=Debug

on a Windows machine. (`-solution=ITK.sln` and `-configuration=Debug` are only from my imagination) But I guess Hans (@hjmjohnson) would know how. Anyhow, let's just proceed with the current PR as it is now anyway!

**Comment by N-Dekker** [comment]
All checks have passed 🎉

**Review by dzenanz — APPROVED** [review]

File: Examples/Segmentation/HoughTransform2DLinesImageFilter.cxx
Comment (N-Dekker) [reply Examples/Segmentation/HoughTransform2DLinesImageFilter.cxx]: @Leengit Would you still want to make a separate PR to address overly verbose type specifiers like `signed long` or `long int` throughout the entire code base? If so, I would like it very likely  😃

**Comment by dzenanz** [comment]
Did this PR introduce [these build errors](https://open.cdash.org/viewBuildError.php?buildid=7831079)?

**Comment by Leengit** [comment]
A fix for those build errors is Pull Request https://github.com/InsightSoftwareConsortium/ITK/pull/3384

**Pull request merged**
